### PR TITLE
Add kubens.service, drop-ins, and kubensenter prefix to kubelet.service

### PR DESCRIPTION
**Pull request by lack**
**- What I did**

In support of openshift enhancement https://github.com/openshift/enhancements/blob/master/enhancements/hide-container-mountpoints.md

This comes in 2 parts:

- One prefixes the kubelet.service files so they start via the new 'kubensenter' script.
- The other adds the new 'kubens.service' and related drop-ins for kubelet and cri-o.

The new 'kubens.service' that drives this new feature is disabled by default, so should be safe to go in as soon as the openshift-hyperkube RPM lands which includes the new 'kubensenter' script required for the 1st part. (https://github.com/openshift/kubernetes/pull/1327)

/hold until openshift-hyperkube arrives with the new kubensenter script.

**- How to verify it**

This cannot be mergedf until the new `kubensenter` script arrives via an updated openshift-hyperkube, but enabling of this feature also requires that a CRI-O change (https://github.com/cri-o/cri-o/pull/5974) that will arrive once RHCOS includes CRI-O 1.25.

Once both prerequisites are in place, this feature should be verified as follows:
- Initially this must have NO EFFECT unless enabled.  Systemd, kubelet, and cri-o are all in the same mount namespace as usual:

```
[root@node ~]# readlink /proc/1/ns/mnt
mnt:[4026531840]
[root@node ~]# readlink /proc/$(pgrep kubelet)/ns/mnt
mnt:[4026531840]
[root@node ~]# readlink /proc/$(pgrep crio)/ns/mnt
mnt:[4026531840]
```
Note all 3 are identical

- Enabling the feature can be done by injecting a single MachineConfig object that enables the new `kubens.service` systemd service.  If this is done, both kubelet and CRI-O should be in a different namespace that systemd, but the system should otherwise behave normally:

With the feature enabled:
```
[root@node ~]# readlink /proc/1/ns/mnt
mnt:[4026531840]
[root@node ~]# readlink /proc/$(pgrep kubelet)/ns/mnt
mnt:[4026534696]
[root@node ~]# readlink /proc/$(pgrep crio)/ns/mnt
mnt:[4026534696]
```
Note that kubelet and crio are identical, but unique from pid 1

**- Description for the changelog**

Adds a new optional systemd service called `kubens.service` which can be enabled to segregate all Kubernetes-specific mount points into a new mount namespace separate from the host OS and systemd.

### CODE REVIEW

**Comment by lack** [comment]
I expect these tests to fail until we get builds with the updated openshift-hyperkube RPM.

**Comment by lack** [comment]
/unhold

The latest RHCOS has the right version of openshift-hyperkube now: https://releases-rhcos-art.cloud.privileged.psi.redhat.com/?stream=releases/rhcos-4.12&release=412.86.202207292203-0#412.86.202207292203-0

**Comment by sinnykumari** [comment]
This PR has changes more specific to kubelet and crio, so will be good to get reviewed by node team first to ensure that it doesn't break anything.

Looking briefly at the changes, don't see any concern from MCO side. Will do another pass once LGTM'ed from node team side.

/assign @rphillips @haircommander

File: vendor/modules.txt [hunk 1]
```diff
@@ -144,6 +144,9 @@ github.com/containers/image/v5/signature
 github.com/containers/image/v5/transports
 github.com/containers/image/v5/types
 github.com/containers/image/v5/version
+# github.com/containers/kubensmnt v1.1.2
+## explicit; go 1.18
+github.com/containers/kubensmnt/utils/systemd
```
Comment (haircommander) [inline vendor/modules.txt:149]: it's not clear to me, why do we need this vendored here?

**Comment by haircommander** [comment]
it may benefit us to put some work into making sure https://bugzilla.redhat.com/show_bug.cgi?id=2057618 is not due to the mount namespace before going forward with this. it's been assigned to me but I've not been able to give it enough attention

File: vendor/modules.txt (hunk 1 above)
Comment (lack) [reply vendor/modules.txt:149]: I'm using my go-embed-vendor trick to get the systemd service (kubens.service) and drop-ins from github.com/containers/kubensmnt to copy in to the templates directory (the copy is done by the `hack/update-templates.sh` script).

Comment (haircommander) [reply vendor/modules.txt:149]: ah right thanks!

**Comment by lack** [comment]
> it may benefit us to put some work into making sure https://bugzilla.redhat.com/show_bug.cgi?id=2057618 is not due to the mount namespace before going forward with this. it's been assigned to me but I've not been able to give it enough attention

I can take a peek at it, to see if I can find it it's related to this feature.

**Comment by lack** [comment]
FYI: CRI-O 1.25 has now landed in RHCOS as of [Build 412.86.202208032219-0](https://releases-rhcos-art.cloud.privileged.psi.redhat.com/?stream=releases/rhcos-4.12&release=412.86.202208032219-0#412.86.202208032219-0)

This means that it should now be possible to spin up a cluster with that RHCOS and this PR and do actual end-to-end testing with the resulting system by enabling the 'kubens.service' in systemd.

**Comment by haircommander** [comment]
/approve

@lack did some investigation trying to see if the bug I attached could be related. Because he did not find anything, and my hunch is only that, I am okay moving forward with this

**Comment by lack** [comment]
I have done some preliminary testing via ClusterBot, building this PR.  I have verified:
- Without any action taken, the cluster comes up fine, and systemd/CRI-O/Kubelet are all in the same namespace
```
sh-4.4# readlink /proc/1/ns/mnt
mnt:[4026531840]
sh-4.4# readlink /proc/$(pgrep kubelet)/ns/mnt
mnt:[4026531840]
sh-4.4# readlink /proc/$(pgrep crio)/ns/mnt
mnt:[4026531840]
```
- When I add a MachineConfig to enable `kubens.service`, the cluster comes up fine, but now with CRI-O and Kubelet in their own mount namespace separate from systemd:
```
sh-4.4# readlink /proc/1/ns/mnt
mnt:[4026531840]
sh-4.4# readlink /proc/$(pgrep kubelet)/ns/mnt
mnt:[4026532278]
sh-4.4# readlink /proc/$(pgrep crio)/ns/mnt
mnt:[4026532278]
```

**Comment by rphillips** [comment]
One sticking point, on RHEL workers if the RPM isn't updated and the new systemd files are laid down, then the kubelet service will fail to start.

**Comment by rphillips** [comment]
/lgtm cancel

^ let's discuss the previous comment

**Comment by lack** [comment]
/hold
@rphillips found a cool edge case, and @sdodson found a cool solution, now implemented in this PR.  I need to take it out for a test drive before we okay this for merge.

**Comment by lack** [comment]
/unhold
Verified the new fix works to toggle the feature on and off, and *should* be safe vs RHEL worker upgrades and any other edge case where this MCO is running against older kubelet version without the new `kubensenter` script.

**Comment by rphillips** [comment]
The changes made resolve my concern with RHEL nodes and lgtm. Tests are failing for some reason which need resolved...

/lgtm

**Comment by cgwalters** [comment]
Side note: pretty sure this would have entirely prevented https://bugzilla.redhat.com/show_bug.cgi?id=2111817

File: templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml [hunk 2]
```diff
@@ -10,6 +10,7 @@ contents: |
 
   [Service]
   Type=notify
+  NotifyAccess=all
```
Comment (cgwalters) [inline templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml:13]: How is this one related?

File: templates/common/_base/units/kubens.service.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  Environment=ENVFILE=%t/kubens/env
+  
+  # Set up the runtime directory as an unbindable mountpoint
+  ExecStartPre=bash -c "findmnt ${RUNTIME_DIRECTORY} || mount --make-unbindable --bind ${RUNTIME_DIRECTORY} ${RUNTIME_DIRECTORY}"
+  # Ensure the bind point exists
+  ExecStartPre=touch ${BIND_POINT}
+  # Use 'unshare' to create the new mountpoint, then 'mount --make-rshared' so it cascades internally
+  ExecStart=unshare --mount=${BIND_POINT} --propagation slave mount --make-rshared /
+  # Finally, set an env pointer for ease-of-use
+  ExecStartPost=awk 'BEGIN {print "KUBENSMNT=${BIND_POINT}" > "${ENVFILE}"}'
```
Comment (cgwalters) [inline templates/common/_base/units/kubens.service.yaml:23]: Why the use of `awk` here?

Comment (lack) [reply templates/common/_base/units/kubens.service.yaml:23]: It's roughly the same as `bash -c 'echo KUBENSMNT=${BIND_POINT} > ${ENVFILE}'` but awk is lighter-weight than bash.

So more of a code-golf "what's better than bash for writing a file from systemd" than anything else.

File: templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml (hunk 2 above)
Comment (lack) [reply templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml:13]: Oh, this is a cool story.

TL;DR="Because we're wrapping kubelet in kubensenter and kubelet is not the direct child of systemd, we need this so kubelet can tell systemd it's running via the `Type=notify` mechanism"

So...

If you see the history of this PR, you can see that we were very consistently failing CI all the way along until the end, and it was very mysterious. Running clusters with the same code, when doing a manual inspection everything looked healthy; `oc get co` showed everything available, `oc get nodes` showed everything Ready.  So the weird CI failures made no sense. For example: CI failed running 'rsh' to cat a file. Manual execution of the same operation always succeeded.  And manually running the MCO e2e tests sometimes passed, or failed in other weird ways intermittently.

The main point of this change is that we want to run `kubelet` inside of the new `kubensenter` wrapper, and due to what turns out to be a short-sighted design choice, that `kubensenter` wrapper does not `exec`, but keeps itself in the PS tree, as an audit that it's there. An important and initially-thought benign impact is that while `kubelet` used to be the direct parent of systemd in the process tree, it is now the child of `kubensenter`.

Now we go deeper into systemd. It turns out that for a `Type=notify` service like this one, unless you say otherwise this `NotifyAccess` defaults to the value `main`, which means that systemd will actively ignore any notification signal that is NOT from the PID directly started by Systemd's ExecStart stanza. Because of this, systemd would start kubelet, let it run for a while, then timeout waiting for the notification signal and restart it.

These periodic restarts would correspond to random CI failures, and were largely invisible to manual checks, especially because when `oc get co` returns everything "available" there's usually no reason to check whether kubelet itself is healthy!  But it turns out, it wasn't...

And now you know the rest of the story.

File: templates/common/_base/units/kubens.service.yaml (hunk 3 above)
Comment (rphillips) [reply templates/common/_base/units/kubens.service.yaml:23]: Is awk in the default install of RH?

Comment (lack) [reply templates/common/_base/units/kubens.service.yaml:23]: I would be shocked to find a linux system that doesn't have awk as part of the default toolset!

Comment (lack) [reply templates/common/_base/units/kubens.service.yaml:23]: Color me shocked!  At least in RHEL7, it turns out gawk is NOT one of the core packages.  I'll reimplement in bash shortly.

Comment (lack) [reply templates/common/_base/units/kubens.service.yaml:23]: Fixed.

File: templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml (hunk 2 above)
Comment (lack) [reply templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml:13]: I have a pending change in kubelet to address this: https://github.com/openshift/kubernetes/pull/1350

Once that's in and kubensenter calls `exec`, we can either edit this PR to remove this `NotifyAccess=all` line, or if this is already merged, I'll launch a follow-up PR to clean this up.

Up to you if you'd prefer to hold this PR on that change, or let this in as-is and fix it later.

Comment (mrunalp) [reply templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml:13]: I think we should make the change to have kubensenter call exec and drop this line.

**Comment by lack** [comment]
The CI checks are going to fail until https://github.com/openshift/kubernetes/pull/1350 goes in and the change makes its way into an RHCOS nightly.  Once that's done I'll remove the hold and retest, and we should be good to go!

File: templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml (hunk 2 above)
Comment (lack) [reply templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml]: I have removed these `NotifyAccess=all` lines.  Once the right kubelet change makes its way through the CI lanes here will pass again and this can be merged :)

File: templates/common/_base/units/crio.service-kubens.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# Autogenerated by hack/update-templates.sh; do not edit
```
Comment (haircommander) [inline templates/common/_base/units/crio.service-kubens.yaml:1]: does this being here mean we'll have the kubens enabled by default? is that intended?

File: templates/worker/01-worker-kubelet/_base/files/kubenswrapper.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# Note: This compatibility wrapper is needed to bridge the gap from OCP 4.11->4.12 when the new 'kubensenter' script is being introduced.
+# It can be removed (and the kubelet.service should call kubensenter directly) when 4.11 is no longer a release we must upgrade from.
+mode: 0755
+path: "/usr/local/bin/kubenswrapper"
```
Comment (haircommander) [inline templates/worker/01-worker-kubelet/_base/files/kubenswrapper.yaml:4]: why /usr/local instead of /usr?

Comment (cgwalters) [reply templates/worker/01-worker-kubelet/_base/files/kubenswrapper.yaml:4]: Ignition can't write to `/usr`.  See https://github.com/openshift/machine-config-operator/issues/3137

**Comment by haircommander** [comment]
LGTM

**Comment by cgwalters** [comment]
From my PoV
/approve
Though one more question - is a machineconfig enabling this going to be added to one of the e2e tests or periodics?

**Comment by sakhoury** [comment]
> From my PoV /approve Though one more question - is a machineconfig enabling this going to be added to one of the e2e tests or periodics?

@cgwalters it is tested in an e2e periodic https://github.com/openshift/release/pull/31715

File: templates/common/_base/units/crio.service-kubens.yaml (hunk 4 above)
Comment (lack) [reply templates/common/_base/units/crio.service-kubens.yaml:1]: This drop-in does nothing except set the environment variable that will be laid down by kubens.service if it's running, so it's safe fro this to be present all the time.

The kubens.service itself is not being enabled by default (see https://github.com/openshift/machine-config-operator/pull/3274/files/d8f42ae7a8e66970a5fcbdaac6fd8712d1466ba7#diff-c38a0be931e5bc991f959879c6af89aee96be94967b4544f63ace1d52e5f88ccR3)
